db/src/excision.rs

```diff
@@ -0,0 +1,129 @@
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::{
+    BTreeSet,
+    BTreeMap,
+};
+
+use mentat_core::{
+    Attribute,
+    Entid,
+    HasSchema,
+    Schema,
+};
+
+use db::{
+    PartitionMapping,
+};
+
+use entids;
+
+use errors::{
+    DbError,
+    Result,
+};
+
+use internal_types::{
+    AEVTrie,
+    filter_aev_to_eav,
+};
+
+use schema::{
+    SchemaBuilding,
+};
+
+use types::{
+    PartitionMap,
+};
+
+/// Details about an excision:
+/// - a target to excise (for now, an entid);
+/// - a possibly empty set of attributes to excise (the empty set means all attributes, not no
+///   attributes);
+/// - and a possibly omitted transaction ID to limit the excision before.  (TODO: check whether
+///   Datomic excises the last retraction before the first remaining assertion, and make our
+///   behaviour agree.)
+///
+/// `:db/before` doesn't make sense globally, since in Mentat, monotonically increasing
+/// transaction IDs don't guarantee monotonically increasing txInstant values.  Therefore, we
+/// accept only `:db/beforeT` and allow consumers to turn `:db/before` timestamps into
+/// transaction IDs in whatever way they see fit.
+#[derive(Clone, Debug, Default)]
+pub(crate) struct Excision {
+    pub(crate) target: Entid,
+    pub(crate) attrs: Option<BTreeSet<Entid>>,
+    pub(crate) before_tx: Option<Entid>,
+}
+
+pub(crate) type Excisions = BTreeMap<Entid, Excision>;
+
+/// Extract excisions from the given transacted datoms.
+pub(crate) fn excisions<'schema>(partition_map: &'schema PartitionMap, schema: &'schema Schema, aev_trie: &AEVTrie<'schema>) -> Result<Option<Excisions>> {
+    let pair = |a: Entid| -> Result<(Entid, &'schema Attribute)> {
+        schema.require_attribute_for_entid(a).map(|attribute| (a, attribute))
+    };
+
+    if aev_trie.contains_key(&pair(entids::DB_EXCISE_BEFORE)?) {
+        bail!(DbError::BadExcision(":db.excise/before".into())); // TODO: more details.
+    }
+
+    let eav_trie = filter_aev_to_eav(aev_trie, |&(a, _)|
+                                     a == entids::DB_EXCISE ||
+                                     a == entids::DB_EXCISE_ATTRS ||
+                                     a == entids::DB_EXCISE_BEFORE_T);
+
+    let mut excisions = BTreeMap::default();
+
+    for (&e, avs) in eav_trie.iter() {
+        for (&(_a, _attribute), ars) in avs {
+            if !ars.retract.is_empty() {
+                bail!(DbError::BadExcision("retraction".into())); // TODO: more details.
+            }
+        }
+
+        let target = avs.get(&pair(entids::DB_EXCISE)?)
+            .and_then(|ars| ars.add.iter().next().cloned())
```

Comment:
Excising multiple targets would require multiple excisions, right? E.g. there's no way to do something like `[{:db/excise 42 43 44}]`?

Reply:
> Excising multiple targets would require multiple excisions, right? E.g. there's no way to do something like [{:db/excise [42 43 44]}]?

No, although I see now that I may have implemented this incorrectly.  `:db/excise` is `:db/cardinality :db.cardinality/many` (which I assume we culted from Datomic), so the expression you give is accepted by the transactors.  I will go add tests to ensure I'm doing reasonable things -- I feel like I might have assumed that the entity corresponds to only one target, which isn't true if this is cardinality many.

Reply:
I did implement this incorrectly.  To capture things correctly I had to make `target` be `targets` and change the underlying materialized view.